From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Fri, 10 Jan 2025 23:47:52 -0600
Subject: [PATCH] Respect build environment

This was largely adapted from Debian, but basically this makes it so that ladspa uses the toolchain set in the build environment and all appropriate flags.

Additionally we specify the PLUGINS_DIR variable so that we can pass in the final path to the ladspa plugin directory so that listplugins works correctly.

---
 src/Makefile | 36 +++++++++++++++++++++---------------
 1 file changed, 21 insertions(+), 15 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index 774930a..98d8959 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -15,10 +15,10 @@ INSTALL_BINARY_DIR	=	/usr/bin/
 
 INCLUDES	=	-I.
 LIBRARIES	=	-ldl -lm -lsndfile
-CFLAGS		=	$(INCLUDES) -Wall -Werror -O2 -fPIC 		\
-			-DDEFAULT_LADSPA_PATH=$(INSTALL_PLUGINS_DIR)
+MANDATORY	=	$(INCLUDES) -fPIC -DDEFAULT_LADSPA_PATH=$(PLUGINS_DIR)
+CFLAGS		?=	-Wall -Werror -O2
 BINFLAGS	=	-fPIE -pie
-CXXFLAGS	=	$(CFLAGS)
+CXXFLAGS	?=	-Wall -Werror -O2
 PLUGINS		=	../plugins/amp.so				\
 			../plugins/delay.so				\
 			../plugins/filter.so				\
@@ -27,8 +27,8 @@ PLUGINS		=	../plugins/amp.so				\
 PROGRAMS	=	../bin/analyseplugin				\
 			../bin/applyplugin 				\
 			../bin/listplugins
-CC		=	cc
-CPP		=	c++
+CC		?=	cc
+CXX		?=	c++
 
 ###############################################################################
 #
@@ -36,25 +36,25 @@ CPP		=	c++
 #
 
 ../plugins/%.so:	plugins/%.c ladspa.h gcc_exports.map
-	$(CC) $(CFLAGS) -o plugins/$*.o -c plugins/$*.c
+	$(CC) $(MANDATORY) $(CFLAGS) -o plugins/$*.o -c plugins/$*.c
 	$(CC)	-o ../plugins/$*.so					\
 		plugins/$*.o						\
 		-shared							\
-		$(CFLAGS)						\
+		$(LDFLAGS)						\
 		-fvisibility=hidden					\
 		-fvisibility-inlines-hidden				\
-		-s							\
 		-Wl,--version-script=gcc_exports.map			
 
 ../plugins/%.so:	plugins/%.cpp ladspa.h gcc_exports.map
-	$(CPP) $(CXXFLAGS) -o plugins/$*.o -c plugins/$*.cpp
-	$(CPP)	-o ../plugins/$*.so					\
+	$(CXX) $(MANDATORY) $(CXXFLAGS) -o plugins/$*.o -c plugins/$*.cpp
+	$(CXX)	-o ../plugins/$*.so					\
 		plugins/$*.o						\
 		-shared							\
-		$(CXXFLAGS)						\
+		$(MANDATORY)					\
+		$(LDFLAGS)						\
+		$(LIBRARIES)						\
 		-fvisibility=hidden					\
 		-fvisibility-inlines-hidden				\
-		-s							\
 		-Wl,--version-script=gcc_exports.map			
 
 ###############################################################################
@@ -105,23 +105,29 @@ targets:	$(PLUGINS) $(PROGRAMS)
 # PROGRAMS
 #
 
+%.o: %.c
+	$(CC) $(CFLAGS) $(MANDATORY) -c -o $@ $< 
+
 ../bin/applyplugin:	applyplugin.o load.o default.o
 	$(CC) $(CFLAGS) $(BINFLAGS)					\
 		-o ../bin/applyplugin					\
 		applyplugin.o load.o default.o				\
-		$(LIBRARIES)
+		$(LIBRARIES)						\
+		$(LDFLAGS)
 
 ../bin/analyseplugin:	analyseplugin.o load.o default.o
 	$(CC) $(CFLAGS) $(BINFLAGS)					\
 		-o ../bin/analyseplugin 				\
 		analyseplugin.o load.o default.o			\
-		$(LIBRARIES)
+		$(LIBRARIES)						\
+		$(LDFLAGS)
 
 ../bin/listplugins:	listplugins.o search.o
 	$(CC) $(CFLAGS) $(BINFLAGS)					\
 		-o ../bin/listplugins	 				\
 		listplugins.o search.o					\
-		$(LIBRARIES)
+		$(LIBRARIES)						\
+		$(LDFLAGS)
 
 ###############################################################################
 #
